jms/src/main/scala/akka/stream/alpakka/jms/JmsSourceStage.scala

```diff
@@ -5,80 +5,319 @@
+    (logic, logic.killSwitch)
+  }
+}
+
+abstract class SourceStageLogic[T](shape: SourceShape[T],
+                                   out: Outlet[T],
+                                   settings: JmsSourceSettings,
+                                   attributes: Attributes)
+    extends GraphStageLogic(shape)
+    with JmsConnector
+    with StageLogging {
+
+  override private[jms] def jmsSettings = settings
+  private val queue = mutable.Queue[T]()
+  private val stopping = new AtomicBoolean(false)
+  private var stopped = false
+
+  private val markStopped = getAsyncCallback[Unit] { _ =>
+    stopped = true
+    if (queue.isEmpty) completeStage()
+  }
+
+  private val markAborted = getAsyncCallback[Throwable] { ex =>
+    stopped = true
+    failStage(ex)
+  }
 
-      override private[jms] def jmsSettings = settings
+  private[jms] def getDispatcher =
+    attributes.get[ActorAttributes.Dispatcher](
+      ActorAttributes.Dispatcher("akka.stream.default-blocking-io-dispatcher")
+    ) match {
+      case ActorAttributes.Dispatcher("") =>
+        ActorAttributes.Dispatcher("akka.stream.default-blocking-io-dispatcher")
+      case d => d
+    }
+
+  private[jms] val handleError = getAsyncCallback[Throwable] { e =>
+    fail(out, e)
+  }
+
+  override def preStart(): Unit = initSessionAsync(getDispatcher)
+
+  private[jms] val handleMessage = getAsyncCallback[T] { msg =>
+    if (isAvailable(out)) {
+      if (queue.isEmpty) {
+        pushMessage(msg)
+      } else {
+        pushMessage(queue.dequeue())
+        queue.enqueue(msg)
+      }
+    } else {
+      queue.enqueue(msg)
+    }
+  }
 
-      private[jms] def getDispatcher =
-        inheritedAttributes.get[ActorAttributes.Dispatcher](
-          ActorAttributes.Dispatcher("akka.stream.default-blocking-io-dispatcher")
-        ) match {
-          case ActorAttributes.Dispatcher("") =>
-            ActorAttributes.Dispatcher("akka.stream.default-blocking-io-dispatcher")
-          case d => d
+  private[jms] def pushMessage(msg: T): Unit
+
+  setHandler(out, new OutHandler {
+    override def onPull(): Unit = {
+      if (queue.nonEmpty) pushMessage(queue.dequeue())
+      if (stopped && queue.isEmpty) completeStage()
+    }
+  })
+
+  private def stopSessions(): Unit =
+    if (stopping.compareAndSet(false, true))
+      Future {
+        try {
+          jmsConnection.stop()
+        } catch {
+          case NonFatal(e) => log.error(e, "Error stopping JMS connection {}", jmsConnection)
         }
+      } andThen {
+        case _ =>
+          val closeSessionFutures = jmsSessions.map { s =>
+            val f = s.closeSessionAsync()
+            f.onFailure { case e => log.error(e, "Error closing jms session") }
+            f
+          }
+          Future.sequence(closeSessionFutures).onComplete { _ =>
+            try {
+              jmsConnection.close()
+              log.info("JMS connection {} closed", jmsConnection)
```

Comment:
What does the toString on jmsConnection return? No secrets or arbitrary long strings?

Reply:
That's a very good point. There are no secrets in case of ActiveMQ, but the same cannot be said for arbitrary messaging systems. It is better to remove this.

Reply:
As there is no evidence of yet there is sensitive information or excessively long strings, I'll leave the error cases in, though. Removing it is easy if anybody finds that is an issue. Please let me know whether you agree.